subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c

```diff
@@ -490,6 +499,13 @@ static int client_transfer(enum coap_method method,
 		(void)k_sem_take(&cb_sem, K_FOREVER); /* Wait for coap_client to exhaust retries */
 	}
 
+	if (!reliable && !err && (user_cb.result_code >= COAP_RESPONSE_CODE_BAD_REQUEST)) {
+		/* NON transfers usually do not use a callback,
+		 * so make sure a bad result is not ignored.
+		 */
+		err = user_cb.result_code;
+	}
+
 	k_sem_give(&serial_sem);
```

Comment:
All the functions in nrf_cloud_coap.c call this function, and there is no other control to prevent a multi-threaded application from calling multiple such functions simultaneously.  We only use a single coap_client instance, and each instance only supports processing one transfer at a time, so we need the semaphore to ensure this.

Reply:
Oh, duh. You know, like, the most basic use-case for semaphores. My bad